File: lib/tailmix/scripting/operations/logical.rb
# frozen_string_literal: true

module Tailmix
  module Scripting
    module Operations
      # Contains logical operations.
      module Logical
        extend self

        def and(interpreter, args)
          interpreter.eval(args[0]) && interpreter.eval(args[1])
        end

        def or(interpreter, args)
          interpreter.eval(args[0]) || interpreter.eval(args[1])
        end

        def not(interpreter, args)
          !interpreter.eval(args[0])
        end

        def if(interpreter, args)
          condition, then_branch, else_branch = args
          result = interpreter.eval(condition)

          branch_to_eval = result ? then_branch : else_branch
          interpreter.eval_branch(branch_to_eval)
        end
      end
    end
  end
end

File: app/javascript/tailmix/runtime/operations/logical.js
// Operations related to logic and control flow.
export const LogicalOperations = {
    and: async (interpreter, args, context) => {
        return (await interpreter.eval(args[0], context)) && (await interpreter.eval(args[1], context));
    },

    or: async (interpreter, args, context) => {
        return (await interpreter.eval(args[0], context)) || (await interpreter.eval(args[1], context));
    },

    not: async (interpreter, args, context) => {
        return !(await interpreter.eval(args[0], context));
    },

    if: async (interpreter, args, context) => {
        const [condition, thenBranch, elseBranch] = args;
        if (await interpreter.eval(condition, context)) {
            await interpreter.run(thenBranch, context);
        } else if (elseBranch) {
            await interpreter.run(elseBranch, context);
        }
    },
};
